state/tx_config.go

```diff
@@ -0,0 +1,188 @@
+	}
+	return cfg.gasPrice
+}
+
+func (cfg *TxConfig) GasLimit() uint64 { return cfg.gas }
+
+func (cfg *TxConfig) KeyName() string { return cfg.keyName }
+
+func (cfg *TxConfig) SignerAddress() string { return cfg.signerAddress }
+
+func (cfg *TxConfig) FeeGranterAddress() string { return cfg.feeGranterAddress }
+
+type jsonTxConfig struct {
+	GasPrice          float64 `json:"gas_price,omitempty"`
+	IsGasPriceSet     bool    `json:"is_gas_price_set,omitempty"`
+	Gas               uint64  `json:"gas,omitempty"`
+	KeyName           string  `json:"key_name,omitempty"`
+	SignerAddress     string  `json:"signer_address,omitempty"`
+	FeeGranterAddress string  `json:"fee_granter_address,omitempty"`
+}
+
+func (cfg *TxConfig) MarshalJSON() ([]byte, error) {
+	jsonOpts := &jsonTxConfig{
+		SignerAddress:     cfg.signerAddress,
+		KeyName:           cfg.keyName,
+		GasPrice:          cfg.gasPrice,
+		IsGasPriceSet:     cfg.isGasPriceSet,
+		Gas:               cfg.gas,
+		FeeGranterAddress: cfg.feeGranterAddress,
+	}
+	return json.Marshal(jsonOpts)
+}
+
+func (cfg *TxConfig) UnmarshalJSON(data []byte) error {
+	var jsonOpts jsonTxConfig
+	err := json.Unmarshal(data, &jsonOpts)
+	if err != nil {
+		return fmt.Errorf("unmarshalling TxConfig: %w", err)
+	}
+
+	cfg.keyName = jsonOpts.KeyName
+	cfg.signerAddress = jsonOpts.SignerAddress
+	cfg.gasPrice = jsonOpts.GasPrice
+	cfg.isGasPriceSet = jsonOpts.IsGasPriceSet
+	cfg.gas = jsonOpts.Gas
+	cfg.feeGranterAddress = jsonOpts.FeeGranterAddress
+	return nil
+}
+
+// estimateGas estimates gas in case it has not been set.
+// NOTE: final result of the estimation will be multiplied by the `gasMultiplier`(1.1) to cover additional costs.
+func estimateGas(ctx context.Context, client *user.TxClient, msg sdktypes.Msg) (uint64, error) {
+	// set fee as 1utia helps to simulate the tx more reliably.
+	gas, err := client.EstimateGas(ctx, []sdktypes.Msg{msg}, user.SetFee(1))
+	if err != nil {
+		return 0, fmt.Errorf("estimating gas: %w", err)
+	}
+	return uint64(float64(gas) * gasMultiplier), nil
+}
+
+// estimateGasForBlobs returns a gas limit as a `user.TxOption` that can be applied to the `MsgPayForBlob` transactions.
+// NOTE: final result of the estimation will be multiplied by the `gasMultiplier`(1.1)
+// to cover additional options of the Tx.
+func estimateGasForBlobs(blobSizes []uint32) uint64 {
+	gas := apptypes.DefaultEstimateGas(blobSizes)
+	return uint64(float64(gas) * gasMultiplier)
+}
+
+func parseAccountKey(kr keyring.Keyring, accountKey string) (sdktypes.AccAddress, error) {
+	rec, err := kr.Key(accountKey)
+	if err != nil {
+		return nil, fmt.Errorf("getting account key: %w", err)
+	}
+	return rec.GetAddress()
+}
+
+func parseAccAddressFromString(addrStr string) (sdktypes.AccAddress, error) {
+	return sdktypes.AccAddressFromBech32(addrStr)
+}
+
+// Attribute is the functional option that is applied to the TxOption instance
+// to configure parameters.
+type Attribute func(cfg *TxConfig)
```

Comment:
If we have TxConfig, we can call it as Option conventially

Reply:
https://github.com/celestiaorg/celestia-node/pull/3349#discussion_r1653022544 ?

Reply:
I am aware @cristaloleg. The reasoning for not calling them Options was that making options for TxOptions is a tautology. Now it's TxConfig, so we should call Attributes as Options, as the original reason doesn't hold anymore.

Reply:
CoreAccessor already has declared Option type in state package, so we should either leave as `Attribute` or rename to `CfgOption`.

Reply:
ConfigOption sgtm